Updated libVC API, fix adi int.

kodaline opened the pull request:
Updated API to newest libVC version. Fixed test adi removing the int
datatype that brings to a floating point exception (caused by division by zero because when having datatype int, integer conversion is used). Code reformat.

skeru commented on tester/main.cpp:
```diff
@@ -301,47 +300,41 @@ int main(int argc, char const *argv[]) {
 		}
 	}
 
-	// compiling and running
-	char* runningArgv[] = {};
-	for (const auto &j : jobList) {
-		std::cout << "working on " << j.first << std::endl;
-		bool ok = j.second->compile();
-		if (!ok) {
-			std::cerr << "Error while compiling " << j.first << std::endl;
-			continue;
-		}
-		main_signature_t* run = reinterpret_cast<main_signature_t*>(j.second->getSymbol());
-		if (run) {
-			run(0, runningArgv);
-		}
-		j.second->fold();
-	}
-
```
I still prefer to keep the configuration loop separated from the compilation loop.

This way, it is more evident the separation of tasks and the capability of vc::Version to store the compilation state.

skeru commented on tester/main.cpp:
```diff
@@ -70,26 +61,34 @@ int main(int argc, char const *argv[]) {
 	vc::Option dataSize_medium("dataType", "-DMEDIUM_DATASET");
 	vc::Option dataSize_large("dataType", "-DLARGE_DATASET");
 	vc::Option dataSize_xlarge("dataType", "-DEXTRALARGE_DATASET");
+
+	const opt_vec_t size_vec = {
+		vc::make_option("-DMINI_DATASET"),
+		vc::make_option("-DSMALL_DATASET"),
+		vc::make_option("-DMEDIUM_DATASET"),
+		vc::make_option("-DLARGE_DATASET"),
+		vc::make_option("-DEXTRALARGE_DATASET"),
+	};
```
When you start using the `vc::make_option()` within the vector initialization, the above variables `datasize_xxx` are no longer needed. I suggest to remove them.

skeru commented on tester/main.cpp:
```diff
@@ -34,34 +37,22 @@ typedef struct cat_t {
-	vc::Option includePolybenchUtilsDir("pb_Dir", "-I", polybench_source + "/utilities");
-	vc::Option optimizationLevel("O","-O","3");
-	vc::Option standardCxx("std","-std=","c++11");
-	std::list<vc::Option> default_options = {
-		standardCxx,
-		includePolybenchUtils,
-		includeCustomUtils,
-		optimizationLevel,
-		includePolybenchUtilsDir,
-		includeCustomUtilsDir,
-	};
```
some of these default options may not be needed anymore, however I still believe the common options should be defined outside loop as they are loop-invariant.

skeru commented on tester/main.cpp:
```diff
@@ -301,47 +300,41 @@ int main(int argc, char const *argv[]) {
 	// preparing versions
 	vc::Version::Builder builder;
 	std::list<run_element_t> jobList;
+	vc::vc_utils_init();
```
This is part of the high-level APIs of libVC. This example should stick to the low-level ones.

skeru commented on tester/main.cpp:
```diff
@@ -301,47 +300,41 @@ int main(int argc, char const *argv[]) {
+						cat.category_folder + "/" +
+						k.kernel_folder;
+					const vc::opt_list_t options = {
+						vc::make_option("-include polybench.c"),
```
the include of the C file should no longer be required.
Instead, there should be the possibility to list more than one file as version source file.
To this end, libVC low level APIs are required (use the builder instead of `vc::createVersion`)

skeru reviewed (changes requested):
Thank you for your work.

I suggested a few changes to your PR to be applied for consistency.

kodaline commented:
updates done following the suggestions.

skeru commented on tester/main.cpp:
```diff
@@ -35,61 +38,58 @@ typedef struct cat_t {
-		includePolybenchUtilsDir,
-		includeCustomUtilsDir,
+	const vc::opt_list_t default_options {
+                vc::make_option("-include utils.hpp"),
```
not sure if leaving a space between `-include` and `utils.hpp` is going to make it work. I remember having troubles in certain configurations. I'm going to double check later.

skeru commented on tester/main.cpp:
```diff
@@ -307,26 +307,26 @@ int main(int argc, char const *argv[]) {
-					builder._functionName = k.function_name;
+						cat.category_folder + "/" +
+						k.kernel_folder;
+					builder._optionList = {
+						vc::make_option("-I" + kernelSourceDir),
+						size.second,
+						datatype.second,
+					};
+					builder._optionList.insert(builder._optionList.end(), default_options.begin(), default_options.end());
+					builder._fileName_src = {kernelSourceDir + "/" + k.fileName, polybench_source + "/utilities/polybench.c"};
+					builder._functionName = {k.function_name};
```
this is definitely OK and should work.

However, libVC version 2 provides additional setters for the builder class, which are a bit more OOP-like.
See, e.g. `vc::Version::Builder::options()`, `vc::Version::Builder::addIncludeDir()` and `vc::Version::Builder::addSourceFile()`.

Can you see if using them would improve the code readability?

skeru commented on tester/main.cpp:
```diff
@@ -307,26 +307,26 @@ int main(int argc, char const *argv[]) {
 			}
 		}
 	}
-
```
this newline feels sad for being removed :cry:

skeru commented on tester/main.cpp:
```diff
@@ -1,5 +1,6 @@
 #include "versioningCompiler/Version.hpp"
-
+#include "versioningCompiler/Utils.hpp"
```
not sure if this include is still required

skeru reviewed (changes requested):
The quality of the PR is definitely improved. I left a few minor comments to be addressed before merging

kodaline commented:
Ok, I added the submodule for libVersioningCompiler so to use the most recent version, updated the cmake so to allow compiling with libclang if available in the system, and update the main.cpp using newest libVC API.

skeru commented on .gitmodules:
```diff
@@ -0,0 +1,4 @@
+[submodule "libVersioningCompiler"]
+	path = libVersioningCompiler
+	url = git@github.com:skeru/libVersioningCompiler.git
+	branch = taffo
```
Mmmmh... i think I know why. Let me rephrase... can we update master on that repo and then link master here?

kodaline replied:
Thanks to the merge of taffo in the libVersioningCompiler repository, now using branch master.

skeru commented on tester/main.cpp:
```diff
@@ -263,7 +262,7 @@ int main(int argc, char const *argv[]) {
 	k_tmp.kernel_folder = "adi";
 	k_tmp.fileName = "adi.c";
-	k_tmp.data_type_option_list = dataType_list_ifd;
+	k_tmp.data_type_option_list = dataType_list_fd;
```
Is this the test you mentioned causing errors with integer configuration?

kodaline replied:
Yup, the int datatype brings to a floating point exception in the adi test case, caused by division by zero because when having datatype int, integer conversion is used.

skeru reviewed (changes requested):
I like the idea of creating a submodule. It definitely makes sense.
Only minor comments left before merging

kodaline commented on CMakeLists.txt:
```diff
@@ -13,9 +13,11 @@ set (DEPS_LIBS ${DEPS_LIBS} ${LIBVC_LIBRARIES})
-	set (CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS} "-DHAVE_CLANG_LIB")
-endif(HAVE_CLANG_LIB)
+if (HAVE_CLANG_LIB_COMPILER)
+        include_directories(${LIBCLANG_INCLUDES})
+        link_directories(${LIBCLANG_LIBRARY_DIR})
```
Yes, the include is not required, but removing also the link, brings to the following error:
```
ld: error: undefined symbol: llvm::EnableABIBreakingChecks
>>> referenced by main.cpp
>>>               CMakeFiles/runPolyBenchSuite.dir/tester/main.cpp.o:(llvm::VerifyEnableABIBreakingChecks)
>>> referenced by ClangLibCompiler.cpp
>>>               ClangLibCompiler.cpp.o:(.data.rel+0x0) in archive ../libVersioningCompiler/build/libVersioningCompiler.a
>>> referenced by FileLogDiagnosticConsumer.cpp
>>>               FileLogDiagnosticConsumer.cpp.o:(.data.rel+0x0) in archive ../libVersioningCompiler/build/libVersioningCompiler.a
>>> referenced by LLVMInstanceManager.cpp
>>>               LLVMInstanceManager.cpp.o:(.data.rel+0x0) in archive ../libVersioningCompiler/build/libVersioningCompiler.a
>>> referenced by TextDiagnostic.cpp
>>>               TextDiagnostic.cpp.o:(.data+0x0) in archive /usr/local/lib/libclangFrontend.a
>>> referenced by DiagnosticRenderer.cpp
>>>               DiagnosticRenderer.cpp.o:(.data+0x0) in archive /usr/local/lib/libclangFrontend.a
>>> referenced by Compilation.cpp
>>>               Compilation.cpp.o:(.data+0x0) in archive /usr/local/lib/libclangDriver.a
>>> referenced by Driver.cpp
>>>               Driver.cpp.o:(.data+0x0) in archive /usr/local/lib/libclangDriver.a
>>> referenced by Action.cpp
>>>               Action.cpp.o:(.data+0x0) in archive /usr/local/lib/libclangDriver.a
>>> referenced by DriverOptions.cpp
>>>               DriverOptions.cpp.o:(.data+0x0) in archive /usr/local/lib/libclangDriver.a
>>> referenced 92 more times

ld: error: undefined symbol: llvm::Error::fatalUncheckedError() const
>>> referenced by Error.h:268 (/home/vagrant/llvm-project/llvm/include/llvm/Support/Error.h:268)
>>>               Darwin.cpp.o:(llvm::Error::assertIsChecked()) in archive /usr/local/lib/libclangDriver.a
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/runPolyBenchSuite.dir/build.make:99: runPolyBenchSuite] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/runPolyBenchSuite.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

skeru replied:
Ok I probably just forgot to add them to the test target.
keep them both

kodaline replied:
Done

kodaline commented on tester/main.cpp:
```diff
@@ -18,6 +18,8 @@ const std::string polybench_source = POLYBENCH_SOURCE_DIRECTORY;
 typedef std::pair<std::string, std::shared_ptr<vc::Version> > run_element_t;
 typedef int (main_signature_t)(int argc, char*argv[]);
+typedef std::vector<vc::Option> opt_vec_t;
+typedef void (*kernel_t)(std::vector<int32_t> &array);
```
Yup, typo

skeru commented on tester/main.cpp:
```diff
@@ -1,5 +1,5 @@
 #include "versioningCompiler/Version.hpp"
-
+#include <vector>
```
If you remove the `opt_vec_t` and replace it with the list-equivalent version, this one should be not required anymore.

kodaline replied:
Changed everything in opt_list_t

kodaline commented on tester/main.cpp:
```diff
@@ -18,6 +18,8 @@ const std::string polybench_source = POLYBENCH_SOURCE_DIRECTORY;
 typedef std::pair<std::string, vc::Option> list_element_t;
 typedef std::pair<std::string, std::shared_ptr<vc::Version> > run_element_t;
 typedef int (main_signature_t)(int argc, char*argv[]);
+typedef std::vector<vc::Option> opt_vec_t;
```
conversion in opt_list_t done
